File: packages/python-sdk/e2b_sdk/session/filesystem_watcher.py
from enum import Enum
from typing import Callable, Set, Optional
from pydantic import BaseModel, PrivateAttr

from e2b_sdk.session.session_connection import SessionConnection
from e2b_sdk.session.filesystem import FilesystemManager


class FilesystemOperation(str, Enum):
    Create = "Create"
    Write = "Write"
    Remove = "Remove"
    Rename = "Rename"
    Chmod = "Chmod"


class FilesystemEvent(BaseModel):
    path: str
    name: str
    operation: FilesystemOperation
    # Unix epoch in nanoseconds
    timestamp: int
    is_dir: bool


FilesystemEventListener = Callable[[FilesystemEvent], None]


class FilesystemWatcher(BaseModel):
    _listeners: Set[FilesystemEventListener] = PrivateAttr(set())
    _rpc_subscription_id: Optional[str] = PrivateAttr(None)

    path: str
    session_connection: SessionConnection

    async def start(self) -> None:
        if self._rpc_subscription_id:
            return

        self._rpc_subscription_id = await self.session_connection.subscribe(
            FilesystemManager.service_name,
            self._handle_filesystem_events,
            "watchDir",
            self.path,
        )

    async def stop(self) -> None:
        self._listeners.clear()
        if self._rpc_subscription_id:
            await self.session_connection.unsubscribe(self._rpc_subscription_id)

    def add_event_listener(self, listener: FilesystemEventListener):
        self._listeners.add(listener)

        def delete_listener() -> None:
            self._listeners.remove(listener)

        return delete_listener

    def _handle_filesystem_events(self, event: FilesystemEvent) -> None:
        for listener in self._listeners:
            listener(event)

File: packages/js-sdk/src/sandbox/filesystemWatcher.ts
import { withTimeout } from '../utils/promise'
import { filesystemService } from './filesystem'
import { CallOpts, SandboxConnection } from './sandboxConnection'

export enum FilesystemOperation {
  Create = 'Create',
  Write = 'Write',
  Remove = 'Remove',
  Rename = 'Rename',
  Chmod = 'Chmod',
}

export interface FilesystemEvent {
  path: string;
  name: string;
  operation: FilesystemOperation;
  // Unix epoch in nanoseconds
  timestamp: number;
  isDir: boolean;
}

export type FilesystemEventListener = (event: FilesystemEvent) => Promise<void> | void;

class FilesystemWatcher {
  // Listeners to filesystem events.
  // Users of the this class can add their listeners to filesystem events
  // via `this.addEventListeners`
  private listeners: Set<FilesystemEventListener>
  private rpcSubscriptionID?: string

  constructor(
    private sessConn: SandboxConnection,
    private path: string,
  ) {
    this.listeners = new Set<FilesystemEventListener>()
  }

  // Starts watching the path that was passed to the contructor
  async start(opts?: CallOpts) {
    const start = async () => {
      // Already started.
      if (this.rpcSubscriptionID) return

      this.handleFilesystemEvents = this.handleFilesystemEvents.bind(this)

      this.rpcSubscriptionID = await this.sessConn._subscribe(
        filesystemService,
        this.handleFilesystemEvents,
        'watchDir',
        this.path,
      )
    }
    return await withTimeout(start, opts?.timeout)()
  }

  // Stops watching the path and removes all listeners.
  async stop() {
    this.listeners.clear()
    if (this.rpcSubscriptionID) {
      await this.sessConn._unsubscribe(this.rpcSubscriptionID)
    }
  }

  addEventListener(l: FilesystemEventListener) {
    this.listeners.add(l)
    return () => this.listeners.delete(l)
  }

  private handleFilesystemEvents(fsChange: FilesystemEvent) {
    this.listeners.forEach((l) => {
      l(fsChange)
    })
  }
}

export default FilesystemWatcher
